controller/appcontroller.go

```diff
@@ -2143,7 +2143,8 @@ func (ctrl *ApplicationController) shouldSelfHeal(app *appv1.Application) (bool,
 		backOff := *ctrl.selfHealBackOff
 		backOff.Steps = int(app.Status.OperationState.Operation.Sync.SelfHealAttemptsCount)
 		var delay time.Duration
-		for backOff.Steps > 0 {
+		steps := backOff.Steps
```

Comment:
Yeah, it happens because 
`next = time.Duration(float64(duration) * factor)
		if cap > 0 && next > cap {
			next = cap
			steps = 0
		}`
Inside backoff library, it is marking steps as 0 and setup next with correct value, and just because of  backOff.Steps > 0 we never reach last iteration. 

Just leaving as explanation for future

Reply:
You still potentially can pass max int value as backoff , but it will not have any performance impact, because here exit in the function if steps are 0

Reply:
Hovewer we can change the loop a bit, but i think it is overcomplicate it without significant benefit 

```
for i := 0; i < steps; i++ {
	delay = backOff.Step()
	
	if i != steps - 1 && backOff.Steps == 0 {
		delay = backOff.Duration
		break
	}
	
}
```